aten/src/ATen/FunctionalTensorImplBase.cpp

```diff
@@ -0,0 +1,184 @@
+        std::shared_ptr<Alias> alias = std::make_shared<Alias>(const_cast<Tensor&>(other));
+        at::ViewMeta noop_view_info(at::ViewMeta::Type::noOp, other.sizes().vec(), other.sizes().vec());
+        // The original tensor wasn't a view - turn it into a (no-op) view.
+        other_impl->alias_ = alias;
+        other_impl->view_metas_ = std::vector<at::ViewMeta>{noop_view_info};
+        // Turn the new tensor into a view too
+        alias_ = alias;
+        view_metas_ = std::vector<at::ViewMeta>{meta};
+    }
+}
+
+bool FunctionalTensorImplBase::is_view() const {
+    return view_metas_.size() != 0;
+}
+
+const char* FunctionalTensorImplBase::tensorimpl_type_name() const {
+    return "FunctionalTensorImplBase";
+}
+
+namespace functionalization {
+namespace impl {
+
+void maybe_add_update(Tensor& self) {
+  auto functional_base_impl = dynamic_cast<at::FunctionalTensorImplBase*>(self.unsafeGetTensorImpl());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(functional_base_impl != nullptr);
+  functional_base_impl->maybe_add_update(self);
+}
+
+void set_view_meta(const at::Tensor& out, const at::Tensor& t, ViewMeta meta) {
+  auto out_impl = dynamic_cast<at::FunctionalTensorImplBase*>(out.unsafeGetTensorImpl());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(out_impl != nullptr);
+  out_impl->set_view_meta(t, meta);
+}
+
+void set_view_meta(const c10::List<at::Tensor>& outs, const at::Tensor& t, ViewMeta meta) {
+  for (auto& out: outs.vec()) {
+    auto out_impl = dynamic_cast<at::FunctionalTensorImplBase*>(out.unsafeGetTensorImpl());
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(out_impl != nullptr);
+    out_impl->set_view_meta(t, meta);
+  }
+}
+
+bool is_alias_tensor(const Tensor& t) {
+  auto t_impl = dynamic_cast<at::FunctionalTensorImplBase*>(t.unsafeGetTensorImpl());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(t_impl != nullptr);
+  return t_impl->is_alias_tensor();
+}
+
+
+bool is_alias_tensor(const c10::List<Tensor>& t_list) {
+  for (auto& t: t_list.vec()) {
+    auto t_impl = dynamic_cast<at::FunctionalTensorImplBase*>(t.unsafeGetTensorImpl());
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(t_impl != nullptr);
+    // There should be an invariant that either all or none of the tensors are alias tensors.
+    // But we don't want to loop over the entire list.
+    return t_impl->is_alias_tensor();
+  }
+  return false;
+}
+
+void maybe_sync(const Tensor& t) {
+  if (t.unsafeGetTensorImpl()->is_wrapped_number()) {
+    // Unfortunately, we can't easily guarantee that wrapped numbers (scalar-tensors)
+    // get wrapped up in a FunctionalTensorWrapper object, since they skip the dispatcher.
+    // That shouldn't matter, since I don't think we're allowed to assign to wrapped numbers anyway.
+    return;
+  }
+  auto functional_impl = dynamic_cast<at::FunctionalTensorImplBase*>(t.unsafeGetTensorImpl());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(functional_impl != nullptr);
+  if (functional_impl->is_view() && !functional_impl->is_up_to_date()) {
+    functional_impl->sync_();
```

Comment:
It seems to me that the `is_view` test should be put in `sync_` and then it is safe to unconditionally all `sync_` at all call sites.